pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1140,11 +1140,46 @@ type WorkflowStep struct {
 	// OnExit is a template reference which is invoked at the end of the
 	// template, irrespective of the success, failure, or error of the
 	// primary template.
+	// DEPRECATED: Use LifecycleHook.template instead.
 	OnExit string `json:"onExit,omitempty" protobuf:"bytes,11,opt,name=onExit"`
+
+	// Hooks holds exit hook which is invoked at the end of the
+	// template, irrespective of the success, failure, or error status of the primary template
+	Hooks *LifecycleHooks `json:"hooks,omitempty" protobuf:"bytes,12,opt,name=hooks"`
+}
+
+type LifecycleHooks struct {
+	// Exit hold the template and arguments which is invoked at the end of the
+	// workflow, irrespective of the success, failure, or error status of the primary template
+	Exit *LifecycleHook `json:"exit," protobuf:"bytes,1,opt,name=exit"`
+}
+
+type LifecycleHook struct {
+	Template  string    `json:"template," protobuf:"bytes,1,opt,name=template"`
+	Arguments Arguments `json:"arguments,omitempty" protobuf:"bytes,2,opt,name=arguments"`
```

Comment:
Normally I'd say `omitempty` types should be pointer types, otherwise they get rendered in the YAML. But I think that `Arguments` is often non-pointer is the code, so I think it is fine to have it non-pointer here.